src/Forms/GridField/GridFieldStateManager.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+
+namespace SilverStripe\Forms\GridField;
+
+use SilverStripe\Control\HTTP;
+use SilverStripe\Control\HTTPRequest;
+
+class GridFieldStateManager implements GridFieldStateManagerInterface
+{
+    /**
+     * @param GridField $gridField
+     * @return string
+     */
+    public function getStateKey(GridField $gridField): string
+    {
+        $i = 0;
+        $form = $gridField->getForm();
+        if ($form) {
+            $controller = $form->getController();
+            while ($controller instanceof GridFieldDetailForm_ItemRequest) {
```

Comment:
This loop is just a convoluted way of counting the number of nested controllers, correct?

Reply:
Yeah, painful, but necessary. GridFields each have their own scope and could in theory you could have two identical grids in the controller stack.